docs/proposals/emptydir-pv.md

```diff
@@ -0,0 +1,151 @@
+
+## AccesModes and Capacity
+
+Capacity on the PV will be ignored till the scheduler can make decisions based on disk. AccessModes will default to `ReadWriteMany`. In the future the kubelet may enforce other AccessModes, but initially they will be ignored (TODO: readonly bind mounts?).
+
+## Implementation
+
+### Overview
+
+A brief overview of the existing volume subsystem on the node:
+```
+                                 updates
+                           asw <- - - - - -
+                            |             |                -----
+    populator -> dsw -> reconciler -> operation executor -| storage plugin      - attach/detach
+                                                          | (emptydir, gce...)  - mount/unmount
+                                                           -----
+```
+
+### Proposed flow
+
+At a high level:
+* Kubelet watches volumes (new)
+* Populator observes pod attached to emptydir PV and adds volume spec to dsw (unchanged)
+* Reconciler notices new volume spec in dsw not in asw, invokes new plugin (unchanged)
+* New plugin attaches, mounts (new)
+* Populator observes pod delete (unchanged)
+* Populator deletes pod from dsw volume map, keeps volume (new, today it deletes volume from map when no pods are using it)
+* Reconciler notices volume in asw with no matching pod in dsw, invokes new plugin (unchanged)
+* New plugin unmounts (new)
+* ...
+* Kubelet notices pv deletion (new)
+* Populator deletes volume from map (new)
+* Reconciler notices volume in asw with no matching volume in dsw, invokes new plugin (unchanged)
+* New plugin detaches (new)
+
+### New storage plugin
+
+We will write an attachable volume plugin that wraps emptydir. The directory is created and deleted in the Attach/Detach stage, BUT these operations are carried out through the same functions that implement emptydir Mount/Unmount. Attach must be idempotent, after creating the `/var/lib/kubelet/volumes/pv-uuid` directory, we record the volume as attached through the node status. Subsequent calls to Attach will no-op. Detach will only get called when the volume is removed from the desired state of the world, which by our design won't happen till the PV is deleted. The permissions on this directory will be the same as emptydir (i.e 777).
+
+This storage plugin will respond "yes" when asked if it owns a volume spec that contains a PV + `sticky-host` annotation, and "no" in every other case. (TODO: do we need to bind mount the directory for every pod?).
+
+## Example
+
+```yaml
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: test
+  annotations:
+    # Set by scheduler or admin
+    volume.alpha.kubernetes.io/sticky-host: ""
+spec:
+  capacity:
+    storage: 10Mi
```

Comment:
Yes, storage on the node will have to be taken into account when
allocating, and it's likely that @vishn's proposals have to be sorted
through to determine how that pool is available.

The node is instance of a dynamic allocator - it's possible nodes might
offer multiple storage classes in a future revision, or that each volume on
the node is counted (i.e. if someone wants to hand out devices per node).

On Fri, Aug 19, 2016 at 8:44 PM, chengyli <notifications@github.com> wrote:

> In docs/proposals/emptydir-pv.md
> <https://github.com/kubernetes/kubernetes/pull/30044#discussion_r75568823>
> :
>
> > +
> > +This storage plugin will respond "yes" when asked if it owns a volume spec that contains a PV + `sticky-host` annotation, and "no" in every other case. (TODO: do we need to bind mount the directory for every pod?).
> > +
> > +## Example
> > +
> > +```yaml
> > +apiVersion: v1
> > +kind: PersistentVolume
> > +metadata:
> > +  name: test
> > +  annotations:
> > +    # Set by scheduler or admin
> > +    volume.alpha.kubernetes.io/sticky-host: ""
> > +spec:
> > +  capacity:
> > +    storage: 10Mi
>
> Will the requests storage size, e.g. storage: 10Mi be considered when
> creating a new emptyDir-pv volume in a node? If yes, then how to know the
> disk allocation status of the node when doing the PV allocation in the
> controller, will there be a new member in NodeStatus?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30044/files/2a9e5a2908a31832129d7cab502b76b41b016069#r75568823>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pxGEB_hFvts9arIO7awyvtN16h7Pks5qhk4BgaJpZM4JcSCY>
> .
>